File: bricks/pac_native_plugin/__brick__/packages/{{name.snakeCase()}}/ios/{{name.snakeCase()}}/Sources/{{name.snakeCase()}}/Platform/{{name.pascalCase()}}HostApiImpl.swift
// Copyright (c) 2026, one of DanhDue ExOICTIF projects. All rights reserved.

import FactoryKit
import Foundation
import UIKit

/// Implementation of Pigeon Host API delegating calls to Clean Architecture UseCases.
public final class {{name.pascalCase()}}HostApiImpl: {{name.pascalCase()}}HostApi, @unchecked Sendable {
    private let getDataUseCase: GetDataUseCase

    public init(getDataUseCase: GetDataUseCase = {{name.pascalCase()}}Container.shared.getDataUseCase()) {
        self.getDataUseCase = getDataUseCase
    }

    public func getPlatformVersion() throws -> String {
        MainActor.assumeIsolated {
            "iOS " + UIDevice.current.systemVersion
        }
    }

    public func getData(completion: @escaping @Sendable (Result<Pigeon{{name.pascalCase()}}Data, Error>) -> Void) {
        Task {
            do {
                let data = try await getDataUseCase.execute()
                let message = Pigeon{{name.pascalCase()}}Data(
                    id: data.id,
                    title: data.title,
                    timestamp: Int64(data.timestamp.timeIntervalSince1970 * 1000)
                )
                completion(.success(message))
            } catch {
                completion(.failure(error))
            }
        }
    }
}

File: bricks/pac_native_plugin/__brick__/packages/{{name.snakeCase()}}/android/src/main/kotlin/com/danhdue/{{name.snakeCase()}}/platform/{{name.pascalCase()}}HostApiImpl.kt
/*
 * Copyright (c) 2026, one of DanhDue ExOICTIF projects. All rights reserved.
 */

package com.danhdue.{{name.snakeCase()}}.platform

import com.danhdue.{{name.snakeCase()}}.domain.usecase.GetDataUseCase
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import javax.inject.Inject

class {{name.pascalCase()}}HostApiImpl @Inject constructor(
    private val getDataUseCase: GetDataUseCase,
    private val coroutineScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
) : {{name.pascalCase()}}HostApi {

    override fun getPlatformVersion(): String {
        return "Android ${android.os.Build.VERSION.RELEASE}"
    }

    override fun getData(callback: (Result<Pigeon{{name.pascalCase()}}Data>) -> Unit) {
        coroutineScope.launch {
            try {
                val outcome = getDataUseCase.execute()
                if (outcome.isSuccess) {
                    val data = outcome.getOrNull()
                    val pigeonData = Pigeon{{name.pascalCase()}}Data(
                        id = data?.id.orEmpty(),
                        title = data?.title.orEmpty(),
                        timestamp = data?.timestamp ?: 0L
                    )
                    callback(Result.success(pigeonData))
                } else {
                    callback(Result.failure(outcome.exceptionOrNull() ?: RuntimeException("Unknown error")))
                }
            } catch (e: Throwable) {
                callback(Result.failure(e))
            }
        }
    }
}
